src/pool.rs

```diff
@@ -603,19 +614,53 @@ impl ConnectionPool {
     /// Get a connection from the pool.
     pub async fn get(
         &self,
-        shard: usize,               // shard number
+        shard: Option<usize>,       // shard number
         role: Option<Role>,         // primary or replica
         client_stats: &ClientStats, // client id
     ) -> Result<(PooledConnection<'_, ServerPool>, Address), Error> {
-        let mut candidates: Vec<&Address> = self.addresses[shard]
-            .iter()
-            .filter(|address| address.role == role)
-            .collect();
+        let mut effective_shard_id: Option<usize> = shard;
+
+        // The base, unsharded case
+        if self.shards() == 1 {
+            effective_shard_id = Some(0);
+        }
+
+        let mut sort_by_error_count = false;
+        let mut candidates: Vec<_> = match effective_shard_id {
+            Some(shard_id) => self.addresses[shard_id].iter().collect(),
+            None => match self.settings.no_shard_specified_behavior {
+                NoShardSpecifiedHandling::Random => self.addresses.iter().flatten().collect(),
+                NoShardSpecifiedHandling::RandomHealthy => {
+                    sort_by_error_count = true;
+                    self.addresses.iter().flatten().collect()
+                }
+                NoShardSpecifiedHandling::Shard(shard) => {
```

Comment:
I'm a bit confused by the naming here. How can `NoShardSpecifiedHandling` then specify the shard it wants the pool to connect to?

Reply:
The way I think about it is that the user can either
- Specify a shard they want to connect to using `SET ...` commands or using the comment or some other query parsing mechanism 
- They specify a specific shard, in which case we need to define a default behavior, this could be either random, healthy_random, or some default shard. 
I am open for suggestions for the name though

Reply:
Maybe `default_shard`?

Reply:
I renamed the config to `default_shard`